lib/python/rose/apps/rose_bunch.py

```diff
@@ -0,0 +1,458 @@
+            for key, proc in procs.items():
+                if proc.poll() is not None:
+                    procs.pop(key)
+                    if proc.returncode:
+                        failed[key] = proc.returncode
+                        app_runner.handle_event(
+                                        FailEvent(key, proc.returncode))
+                        if self.dao:
+                            self.dao.update_command_state(key, self.dao.S_FAIL)
+                        if self.FAIL_HANDLE == self.TYPE_ABORT_ON_FAIL:
+                            abort = True
+                            app_runner.handle_event(AbortEvent())
+                    else:
+                        app_runner.handle_event(
+                                    SucceededEvent(key, proc.returncode))
+                        if self.dao:
+                            self.dao.update_command_state(key, self.dao.S_PASS)
+
+            while len(procs) < self.MAX_PROCS and commands and not abort:
+                key = self.invocation_names[0]
+                command = commands.pop(key)
+                self.invocation_names.pop(0)
+                cmd = command.get_command()
+                cmd_stdout = log_format % command.get_out_file()
+                cmd_stderr = log_format % command.get_err_file()
+
+                if self.dao:
+                    if self.dao.check_has_succeeded(key):
+                        app_runner.handle_event(SkippingEvent(key))
+                        continue
+                    else:
+                        self.dao.add_command(key)
+
+                app_runner.handle_event(LaunchEvent(key, cmd))
+                procs[key] = app_runner.popen.run_bg(
+                                                cmd,
+                                                shell=True,
+                                                stdout=open(cmd_stdout, 'w'),
+                                                stderr=open(cmd_stderr, 'w'))
+            sleep(self.SLEEP_DURATION)
+
+        if self.dao:
+            self.dao.close()
+
+        if failed:
+            return 1
+        else:
+            return 0
+
+
+class RoseBunchCmd(object):
+
+    """A command instance to run."""
+
+    OUTPUT_TEMPLATE = "bunch.%s.%s"
+
+    def __init__(self, name, command, index):
+        self.command_format = command
+        self.argsdict = {}
+        self.name = name
+        self.index = index
+
+
+    def get_command(self):
+        """Return the command that will be run"""
+
+        return self.command_format % self.argsdict
+
+    def get_out_file(self):
+        """Return output file name"""
+
+        return self.OUTPUT_TEMPLATE % (self.name, "out")
+
+    def get_err_file(self):
+        """Return error file name"""
+
+        return self.OUTPUT_TEMPLATE % (self.name, "err")
+
+
+class RoseBunchDAO(object):
+
+    """Database object for rose_bunch"""
+
+    T_COMMANDS = "commands"
+    T_CONFIG = "config"
```

Comment:
@benfitzpatrick will probably argue that we ought to say `TABLE` something or something `TABLE`.